src/util/http/HttpClient.h

```diff
@@ -58,13 +58,21 @@ class HttpClientImpl {
       std::string_view target, ad_utility::SharedCancellationHandle handle,
       std::string_view requestBody = "",
       std::string_view contentTypeHeader = "text/plain",
-      std::string_view acceptHeader = "text/plain");
+      std::string_view acceptHeader = "text/plain",
+      const std::unordered_map<std::string_view, std::string_view>&
+          customHeaders = {});
 
   // Simple way to establish a websocket connection
   boost::beast::http::response<boost::beast::http::string_body>
   sendWebSocketHandshake(const boost::beast::http::verb& method,
                          std::string_view host, std::string_view target);
 
+  // Upgrade the client to a websocket connection and return a generator of
+  // messages received from the server.
+  static cppcoro::generator<std::string> readWebSocketStream(
+      std::unique_ptr<HttpClientImpl> client, std::string_view host,
+      std::string_view target);
+
```

Comment:
We are trying to get away from `generator`s, 
Please have a look at the `InputRangeTypeErased`  etc. classes at the end of `Iterators.h` , and ask me if you have trouble.
And as stated above, you should have a possibility to cancel the websocket stream if the outer operation is cancelled.